shadowsocks/tcp.go

```diff
@@ -106,24 +106,35 @@ func findAccessKey(clientConn onet.DuplexConn, cipherList CipherList) (string, o
 		}
 		// Move the active cipher to the front, so that the search is quicker next time.
 		cipherList.SafeMarkUsedByClientIP(entry, clientIP)
-		ssr := NewShadowsocksReader(io.MultiReader(bytes.NewReader(replayBytes), clientConn), cipher)
+		ssr := NewShadowsocksReader(io.MultiReader(bytes.NewReader(firstBytes), clientConn), cipher)
 		ssw := NewShadowsocksWriter(clientConn, cipher)
-		return id, onet.WrapConn(clientConn, ssr, ssw).(onet.DuplexConn), nil
+		return id, onet.WrapConn(clientConn, ssr, ssw).(onet.DuplexConn), salt, nil
 	}
-	return "", clientConn, fmt.Errorf("Could not find valid TCP cipher")
+	return "", clientConn, firstBytes, fmt.Errorf("Could not find valid TCP cipher")
 }
 
 type tcpService struct {
-	listener    *net.TCPListener
+	listener *net.TCPListener
+	// `ciphers` is a pointer to SSPort.cipherList, which can be updated by SSServer.loadConfig.
 	ciphers     *CipherList
 	m           metrics.ShadowsocksMetrics
 	isRunning   bool
 	readTimeout time.Duration
+	replayCache ReplayCache
 }
 
+// Prevent replays of this many of the most recent handshakes.
```

Comment:
Might be worth commenting on the probability of collision somewhere.

Reply:
Done